pkg/protocol/engine/booker/inmemorybooker/booker.go

```diff
@@ -132,15 +115,38 @@ func (b *Booker) Queue(block *blocks.Block) error {
 
 func (b *Booker) Shutdown() {
 	b.TriggerStopped()
-	b.workers.Shutdown()
 }
 
-func (b *Booker) setRetainBlockFailureFunc(retainBlockFailure func(iotago.BlockID, apimodels.BlockFailureReason)) {
-	b.retainBlockFailure = retainBlockFailure
+func (b *Booker) setupBlock(block *blocks.Block) {
+	var unbookedParentsCount atomic.Int32
+	unbookedParentsCount.Store(int32(len(block.Parents())))
+
+	block.ForEachParent(func(parent iotago.Parent) {
+		parentBlock, exists := b.blockCache.Block(parent.ID)
+		if !exists {
+			panic("cannot setup block without existing parent")
```

Comment:
We really need to start thinking about a proper way to log these things. Panicing everywhere will just increase the likelihood of the node going down even though it doesn't have to. Simply logging and ignoring this block should be enough

Reply:
Passing the error to the errorHandler and skipping this parent.
It would be better to bail the setup for the whole block and undo the hooking we did so far, but it is tricky and ugly; let's rely instead on the GC to eventually cleanup everything, as this block can never become solid.